File: pymath/algebra.py
def solve_slae(A, b):
    """Solve a system of linear equations using Gaussian elimination with partial pivoting."""
    n = len(A)
    for i in range(n):
        # Partial pivoting: find the maximum element in the current column
        max_row = max(range(i, n), key=lambda r: abs(A[r][i]))
        if A[max_row][i] == 0:
            raise ValueError("Singular matrix: system has no unique solution")

        # Swap rows if needed
        if max_row != i:
            A[i], A[max_row] = A[max_row], A[i]
            b[i], b[max_row] = b[max_row], b[i]

        # Gaussian elimination
        for j in range(i + 1, n):
            factor = A[j][i] / A[i][i]
            for k in range(i, n):
                A[j][k] -= factor * A[i][k]
            b[j] -= factor * b[i]

    # Back substitution
    x = [0] * n
    for i in range(n - 1, -1, -1):
        if A[i][i] == 0:
            raise ValueError("Singular matrix detected during back substitution")
        x[i] = (b[i] - sum(A[i][j] * x[j] for j in range(i + 1, n))) / A[i][i]
    return x

File: src/algebra.cpp
#include "algebra.hpp"
#include <cmath> 

namespace algebra {
    std::vector<double> solve_slae(const std::vector<std::vector<double>>& A, const std::vector<double>& b) {
        size_t n = A.size();
        std::vector<std::vector<double>> matrix = A;
        std::vector<double> result = b;

        for (size_t i = 0; i < n; ++i) {
            size_t max_row = i;
            for (size_t k = i + 1; k < n; ++k) {
                if (std::abs(matrix[k][i]) > std::abs(matrix[max_row][i])) {
                    max_row = k;
                }
            }

            if (std::abs(matrix[max_row][i]) < 1e-10) {
                throw std::invalid_argument("Singular matrix detected");
            }

            std::swap(matrix[i], matrix[max_row]);
            std::swap(result[i], result[max_row]);

            for (size_t k = i + 1; k < n; ++k) {
                double factor = matrix[k][i] / matrix[i][i];
                result[k] -= factor * result[i];
                for (size_t j = i; j < n; ++j) {
                    matrix[k][j] -= factor * matrix[i][j];
                }
            }
        }

        std::vector<double> x(n);
        for (int i = n - 1; i >= 0; --i) {
            double sum = 0.0;
            for (size_t j = i + 1; j < n; ++j) {
                sum += matrix[i][j] * x[j];
            }
            x[i] = (result[i] - sum) / matrix[i][i];
        }

        return x;
    }
}